### PR TITLE
[Master] Fix resolve entity name to not dive inside property access expression when the expression is not entity name

### PR DESCRIPTION
**Pull request by yuit**
While writing tests for dynamic import, I ran into `Debug.fail("Unknown entity name kind.");` inside `resolveEntityName` function, turned out that it not only causes failure when writing out ".types" files but also disable some language services features (see test files)

### CODE REVIEW

File: tests/cases/fourslash/findAllReferPropertyAccessExpressionHeritageClause.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+////     B : CTor
+//// };
+//// class B {}
+//// function foo() {
+////     return {[|B|]: B};
+//// }
+//// class C extends (foo()).[|B|] {}
+//// class C1 extends foo().[|B|] {}
+
+const [def, ref1, ref2] = test.ranges();
```
Comment (sandersn) [inline tests/cases/fourslash/findAllReferPropertyAccessExpressionHeritageClause.ts]: Try this instead:

```ts
const rs = test.ranges();
for (const r of rs) {
  verify.referencesOf(r, rs));
}
```

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -1511,6 +1525,15 @@ namespace ts {
                     return undefined;
                 }
             }
+            else if (name.kind === SyntaxKind.ParenthesizedExpression) {
+                // If the expression in parenthsizedExpression is not an entity-name (e.g. it is a call expression), it won't be able to successfully resolve the name.
```
Comment (sandersn) [inline src/compiler/checker.ts]: typo:parenthesized

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -1511,6 +1525,15 @@ namespace ts {
...
                 }
             }
+            else if (name.kind === SyntaxKind.ParenthesizedExpression) {
+                // If the expression in parenthsizedExpression is not an entity-name (e.g. it is a call expression), it won't be able to successfully resolve the name.
+                // This is the case when we are trying to do any language service operation in heritage clauses. By return undefined, the getSymbolOfEntityNameOrPropertyAccessExpression
+                // will attempt to checkPropertyAccessExpression to resolve symbol.
+                // i.e class C extends foo()./*do language service operation here*/B {}
+                return isEntityNameExpression(name.expression) ?
+                    resolveEntityName(name.expression as EntityNameOrEntityNameExpression, meaning, ignoreErrors, dontResolveAlias, location) :
+                    undefined;
```
Comment (sandersn) [inline src/compiler/checker.ts]: should we also give an error in the undefined case?

File: tests/cases/conformance/es6/classDeclaration/emitClassDeclarationWithPropertyAccessInHeritageClause1.ts [hunk 4]
```diff
@@ -0,0 +1,12 @@
+interface I {}
+interface CTor {
+    new (hour: number, minute: number): I
+}
+var x: {
+    B : CTor
+};
```
Comment (sandersn) [inline tests/cases/conformance/es6/classDeclaration/emitClassDeclarationWithPropertyAccessInHeritageClause1.ts]: turns out lines 1-7 aren't used here either!

File: src/compiler/checker.ts (hunk 3 above)
Comment (yuit) [inline src/compiler/checker.ts]: We shouldn't because the caller will already handle this case.

**Pull request merged**